### ISSUE TITLE
False positive: decorated "unused" function

### ISSUE THREAD

**Issue by smurfix** [PyCQA/pylint#4254]
### Steps to reproduce

Given a file `a.py`:
```python
# pylint: disable=C
def foo():
    _dispatch = {}

    def save(fn):
        _dispatch[fn.__name__] = fn
        return fn

    @save
    def bar():
        ...
    # do whatever
foo()
```

### Current behavior

```
/tmp/a.py:10:4: W0612: Unused variable 'bar' (unused-variable)
```

### Expected behavior

No warning.

### pylint --version output

Result of `pylint --version` output:
```
pylint 2.7.2
astroid 2.5.1
Python 3.9.1+ (default, Feb  5 2021, 13:46:56) 
[GCC 10.2.1 20210110]
```

**Comment by elizabethbott** [comment]
Working on this, pull request to follow in next few weeks.
